[BEAM-22] Correct sequence of pending element updates in InMemoryWatermarkManager

tgroh opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

Adding additional pending elements/timers (and thus holds) always comes
before removing existing holds, so the watermark is never observed under
more stringent restrictions than currently exist.

tgroh commented:
R: @bjchambers

bjchambers commented:
The description is a bit unclear -- please update. Also, this should probably have a test to verify proper behavior.

tgroh commented:
Fixed description, added comment.

The test is pretty complex, and I'm going to replace the update implementation with one that has a dedicated thread, and only on #extractFiredTimers; which will cause the test to no longer be particularly relevant.
